### PR TITLE
DATAUP-314: job log viewer flotsam and jetsam

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

Standardised the code that generates the job status indicator in a minimised cell so it lives in the Jobs module. At the same time, I got rid of a ton of unnecessary divs from the cell header and generally slimmed things down a lot.

See inline comments for more info.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-314
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to your favourite bulk upload cell to see the difference!

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/scss/modules/_colours.scss [hunk 1]
```diff
@@ -126,6 +118,20 @@ $palette: (
...
+  "estimating": map-get($kbase-palette, 'info'),
+  "queued": map-get($kbase-palette, 'info'),
+  "running": map-get($kbase-palette, 'info'),
+  "completed": map-get($kbase-palette, 'success-dark'),
+  "terminated": map-get($kbase-palette, 'warning-dark'),
+  "error": map-get($kbase-palette, 'error'),
+  "does_not_exist": map-get($kbase-palette, 'error'),
+);
+
+$palette: map-merge($kbase-palette, $job_status_colours);
```
Comment (ialarmedalien) [inline kbase-extension/scss/modules/_colours.scss:133]: create a new palette using the job status colours and the existing kbase colours

File: kbase-extension/scss/partials/_cellToolbar.scss [hunk 2]
```diff
@@ -0,0 +1,132 @@
+$default-icon-size: 56px;
+
+.kb-cell-toolbar {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_cellToolbar.scss:3]: Moved a ton of css out of the JS file and into its rightful home in a SCSS file.

File: kbase-extension/scss/partials/_custom.scss [hunk 3]
```diff
@@ -304,16 +304,6 @@ div {
     .btn-default {
       color: rgb(204 204 204);
     }
-
-    .kb-cell-toolbar {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_custom.scss:308]: now in the `_cellToolbar` file

File: kbase-extension/scss/partials/_job-status.scss [hunk 4]
```diff
@@ -134,91 +133,52 @@
...
   &__summary {
     color: use-color('ink');
     font-weight: 700;
+  }
 
-    /* .kb-job-status__summary--completed */
-    &--completed {
-      color: use-color('success-dark');
-      font-weight: 700;
+  @each $key, $value in $job_status_colours {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_job-status.scss:162]: programmatically generate a set of classes for each job state with the appropriate text colour.

File: kbase-extension/scss/partials/_narrative.scss [hunk 5]
```diff
@@ -1922,37 +1922,6 @@ div#notebook_panel { /* FIXME Jupyter css */
   font-style: italic;
 }
 
-.kb-cell-toolbar {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_narrative.scss:1925]: now refactored and in the `_cellToolbar.scss` partial

File: kbase-extension/static/kbase/custom/custom.js [hunk 6]
```diff
@@ -532,22 +532,13 @@ define([
             $cellNode.find('.inner_cell > div:nth-child(2)').addClass('hidden');
             $cellNode.find('.inner_cell > div:nth-child(3)').addClass('hidden');
             utils.setCellMeta(this, 'kbase.cellState.showTitle', true);
-            utils.setCellMeta(this, 'kbase.cellState.message', '', true);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/custom/custom.js:535]: only place where this is mentioned ==> get rid of it

File: kbase-extension/static/kbase/custom/custom.js [hunk 7]
```diff
@@ -532,22 +532,13 @@ define([
...
         };
 
         p.maximize = function () {
             const $cellNode = $(this.element);
             $cellNode.find('.inner_cell > div:nth-child(2)').removeClass('hidden');
             $cellNode.find('.inner_cell > div:nth-child(3)').removeClass('hidden');
             utils.setCellMeta(this, 'kbase.cellState.showTitle', false);
-
-            // this is a little distracting to see in all markdown cells at all times.
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/custom/custom.js:544]: out of date comment

File: kbase-extension/static/kbase/js/widgets/narrative_core/ipythonCellMenu.js [hunk 8]
```diff
@@ -1,19 +1,10 @@
...
  */
-define([
-    'jquery',
-    'notebook/js/celltoolbar',
-    'base/js/namespace',
-    'kb_common/html',
-    // 'kbaseNarrativeCellMenu',
-    'kbaseCellToolbarMenu',
-], (
+define(['jquery', 'notebook/js/celltoolbar', 'common/html', 'kbaseCellToolbarMenu'], (
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/ipythonCellMenu.js:4]: eslint changes

File: kbase-extension/static/kbase/js/widgets/narrative_core/ipythonCellMenu.js [hunk 9]
```diff
@@ -40,18 +31,18 @@ define([
     }
 
     function status(toolbarDiv, cell) {
-        const status = getMeta(cell, 'attributes', 'status'),
-            content = span({ style: { fontWeight: 'bold' } }, status);
+        const _status = getMeta(cell, 'attributes', 'status'),
+            content = span({ style: { fontWeight: 'bold' } }, _status);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/ipythonCellMenu.js:35]: I am not sure that these functions are actually used anywhere...

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 10]
```diff
@@ -1,24 +1,23 @@
...
     'jquery',
-    'kb_common/html',
+    'common/html',
     'common/events',
     'base/js/namespace',
     'common/utils',
-    'util/bootstrapDialog',
-    'kbase/js/widgets/appInfoPanel',
     'narrativeConfig',
+    'common/jobs',
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:8]: Ran eslint on the file, removed quite a bit of cruft, and then refactored the cell toolbar to remove a load of unneeded DIVs.

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 11]
```diff
@@ -50,52 +48,25 @@ define([
...
-                showSubtitle = utils.getCellMeta(cell, 'kbase.cellState.showSubtitle', true);
-
-            if (showTitle) {
-                return subTitle;
-            }
-            return '';
+            return showTitle ? title : '';
         }
 
-        function getCellInfoLink(cell, events) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:74]: this is not used anywhere

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 12]
```diff
@@ -105,163 +76,93 @@ define([
...
         }
 
-        function isCodeShowing(cell) {
-            if (cell.isCodeShowing) {
-                return cell.isCodeShowing();
-            }
-            return null;
-        }
-
-        function doShowInfoModal(e) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:115]: not used

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 13]
```diff
@@ -293,288 +194,185 @@ define([
...
-            }
-            return span({
-                class: 'fa fa-thumbs-down fa-2x',
-                style: {
-                    verticalAlign: 'top',
-                },
-            });
-        }
-
-        function minimizedStatus(mode, stage) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:349]: now in `Jobs.createJobStatusFromFsm`

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 14]
```diff
@@ -293,288 +194,185 @@ define([
...
+                                            class: `${cssBaseClass}__icon--${toggleIcon} fa fa-${toggleIcon}-square-o fa-lg`,
+                                        }),
+                                    ]
+                                );
+                            })(),
                         ]
                     ),
-                ]);
+                ],
+                content = div(
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:325]: much simpler cell layout

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 15]
```diff
@@ -105,163 +76,93 @@ define([
...
-                    icon: {
-                        type: 'table',
-                        color: 'black',
-                    },
-                    id: events.addEvent({
-                        type: 'click',
-                        handler: () => {
-                            cell.toggleBatch();
+            if (!readOnly) {
+                if (_cell.toggleBatch && Config.get('features').batchAppMode) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:139]: don't put in the batch option if the cell is read-only

File: test/unit/spec/narrative_core/kbaseCellToolbarMenu-spec.js [hunk 16]
```diff
@@ -1,105 +1,382 @@
...
-                metadata: {
-                    kbase: {
-                        appCell: {
-                            fsm: fsm,
-                        },
-                        cellState: {
-                            toggleMinMax: collapsedState,
-                        },
-                        type: 'app',
+    function generateCell(mode, stage, collapsedState) {
```
Comment (ialarmedalien) [inline test/unit/spec/narrative_core/kbaseCellToolbarMenu-spec.js:8]: Added in tests for most of the cellToolbar functionality.

File: Gruntfile.js [hunk 17]
```diff
@@ -44,11 +44,7 @@ module.exports = function (grunt) {
                         ]),
                     ],
                 },
-                src: [
```
Comment (ialarmedalien) [inline Gruntfile.js:47]: I thought that this had been fixed ages ago. Hmm.

File: kbase-extension/scss/partials/_job-status.scss [hunk 18]
```diff
@@ -134,91 +133,52 @@
...
+      color: use-color($key);
       font-weight: 700;
+      font-size: 16px;
+      text-transform: capitalize;
+      margin: 0 0.5rem;
     }
   }
-
-  /* .kb-job-status__param_list */
-  &__param_list {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_job-status.scss:211]: moved up ^^^

File: kbase-extension/static/kbase/js/common/cellComponents/tabs/jobStatus/jobStateListRow.js [hunk 19]
```diff
@@ -28,7 +28,7 @@ define([
                     class: `fa fa-circle ${cssBaseClass}__icon--${jobStateObject.status}`,
                 }) +
                 ' ' +
-                Jobs.jobLabel(jobStateObject)
+                Jobs.jobLabel(jobStateObject, true)
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/cellComponents/tabs/jobStatus/jobStateListRow.js:31]: show the error message if it's an error status job state

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 20]
```diff
@@ -50,52 +48,25 @@ define([
...
-            }
-            return '';
+            return showTitle ? subTitle : '';
         }
 
         function doToggleMinMaxCell(e) {
             if (e.getModifierState) {
                 const modifier = e.getModifierState('Alt');
                 if (modifier) {
-                    console.log('I want to toggle all cells!');
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:98]: yeah, helpful 😖

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js [hunk 21]
```diff
@@ -293,288 +194,182 @@ define([
...
+                                    )
+                                );
                             })
                             .join(''),
                     ]
                 ),
             ]);
         }
 
-        function buildIcon(cell) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseCellToolbarMenu.js:337]: moved up in the file

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeCellMenu.js [hunk 22]
```diff
@@ -357,7 +357,9 @@ define([
                         .append(
                             $('<div class="col-sm-8">')
                                 .append(
-                                    $('<div class="buttons pull-left">').append(this.$collapseBtn)
+                                    $(
+                                        '<div class="kb-cell-toolbar__buttons-container buttons pull-left">'
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeCellMenu.js:361]: update the class names -- not sure if this module is used anywhere, but if it is, the classname is still correct.

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeWorkspace.js [hunk 23]
```diff
@@ -384,7 +384,7 @@ define([
                 '#kb-share-btn',
                 '#kb-save-btn', // action btns
                 '#kb-ipy-menu', // kernel menu
-                '.kb-cell-toolbar .buttons.pull-right', // Jupyter icons
+                '.kb-cell-toolbar__buttons-container', // Jupyter icons
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeWorkspace.js:387]: ditto - class name update

**Review by briehl — APPROVED** [review]
I think this got out of sync with the `DATAUP-314_job_log_viewer_II` branch as there's some duplicated code changes there. Or Github just really wants me to review it multiple times. I'm assuming that jobLogViewer and its tests haven't changed between the two PRs.

**Pull request merged**